ddtrace/internal/_threads.cpp

```diff
@@ -0,0 +1,528 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+
+#include "structmember.h"
+
+#include <stddef.h>
+
+#include <chrono>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <thread>
+
+// ----------------------------------------------------------------------------
+/**
+ * Ensure that the GIL is held.
+ */
+class GILGuard
+{
+  public:
+    inline GILGuard()
+    {
+        if (!_Py_IsFinalizing())
+            _state = PyGILState_Ensure();
+    }
+    inline ~GILGuard()
+    {
+        if (PyGILState_Check())
+            PyGILState_Release(_state);
+    }
+
+  private:
+    PyGILState_STATE _state;
+};
+
+// ----------------------------------------------------------------------------
+/**
+ * Release the GIL to allow other threads to run.
+ */
+class AllowThreads
+{
+  public:
+    inline AllowThreads()
+    {
+        if (!_Py_IsFinalizing())
+            _state = PyEval_SaveThread();
+    }
+    inline ~AllowThreads()
+    {
+        if (!_Py_IsFinalizing())
+            PyEval_RestoreThread(_state);
+    }
+
+  private:
+    PyThreadState* _state;
+};
+
+// ----------------------------------------------------------------------------
+class PyRef
+{
+  public:
+    inline PyRef(PyObject* obj)
+      : _obj(obj)
+    {
+        Py_INCREF(_obj);
+    }
+    inline ~PyRef() { Py_DECREF(_obj); }
+
+  private:
+    PyObject* _obj;
+};
+
+// ----------------------------------------------------------------------------
+class Event
+{
+  public:
+    void set()
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+        _set = true;
+        _cond.notify_all();
+    }
+
+    void wait()
+    {
+        std::unique_lock<std::mutex> lock(_mutex);
+        _cond.wait(lock, [this]() { return _set; });
+    }
+
+    bool wait(std::chrono::milliseconds timeout)
+    {
+        std::unique_lock<std::mutex> lock(_mutex);
+        return _cond.wait_for(lock, timeout, [this]() { return _set; });
+    }
+
+    void clear()
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+        _set = false;
+    }
+
+  private:
+    std::condition_variable _cond;
+    std::mutex _mutex;
+    bool _set = false;
+};
+
+// ----------------------------------------------------------------------------
+typedef struct periodic_thread
+{
+    PyObject_HEAD
+
+      double interval;
```

Comment:
```suggestion
    double interval;
```

Reply:
🤔 I don't know if the formatter would bark at us for this. This is how my IDE has decided to format this line 🤷

Reply:
oh, maybe because the `PyObject_HEAD` above with no `;`? might have it confused?